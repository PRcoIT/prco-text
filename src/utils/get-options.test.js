const getOptions = require("./get-options").getOptions;

describe("get-options", () => {
  let response;
  let commandOptions;

  const getCommandOptions = (options) => {
    const commandOptions = {
      language: "node",
      command: "prco-check-status",
      from: "415-935-3327",
      to: "415-935-3327",
      message: "hi",
    };

    Object.keys(options).forEach((option) => {
      commandOptions[option] = options[option];
    });

    const result = [
      commandOptions["language"],
      commandOptions["command"],
      "--from",
      commandOptions["from"],
      "--to",
      commandOptions["to"],
      "--message",
      commandOptions["message"],
    ];

    return result;
  };

  beforeEach(() => {
    response = "";
  });

  describe("invalid options", () => {
    describe("when command given no options", () => {
      it("should produce error message and show usage", async () => {
        process.argv = ["node", "prco-check-status"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when config_env_file not found", () => {
      it("should show error and usage", async () => {
        process.argv = ["node", "prco-check-status", "-c /foo"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when command given invalid service option (not wis or oneguard)", () => {
      it("should show error and usage", async () => {
        process.argv = ["node", "prco-check-status", "-s foo"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when command given invalid 'from' option", () => {
      it("should show error and usage", async () => {
        process.argv = ["node", "prco-check-status"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });

        process.argv = ["node", "prco-check-status", "-f 555"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when comand given invalid 'to' option", () => {
      it("should show error and usage", async () => {
        process.argv = ["node", "prco-check-status"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });

        process.argv = ["node", "prco-check-status", "-t 555"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when command given invalid 'message' option", () => {
      it("should show error and usage", async () => {
        process.argv = ["node", "prco-check-status"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });
  });

  describe("phone number transformation and validation", () => {
    describe("when command given 'from' with dashes", () => {
      it("should modify 'from' field and remove dashes and add usa country code", async () => {
        process.argv = getCommandOptions({ from: "415-935-3327", to: "415935" });
        getOptions()
          .then(console.log)
          .catch(async (e) => {
            expect(e).toMatchSnapshot();
          });
      });
    });

    describe("when command given 'from' missing country code", () => {
      it("should modify 'from' field and add usa country code", async () => {
        process.argv = getCommandOptions({ from: "4159353327", to: "415935" });
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when command given invalid 'from' number", () => {
      it("should show error and usage", async () => {
        process.argv = getCommandOptions({ from: "4159353", to: "4159353327" });
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when command given 'to' with dashes", () => {
      it("should modify 'to' field and remove dashes and add usa country code", async () => {
        process.argv = getCommandOptions({ from: "415935", to: "415-935-3327" });
        getOptions()
          .then(console.log)
          .catch(async (e) => {
            expect(e).toMatchSnapshot();
          });
      });
    });

    describe("when command given 'to' missing country code", () => {
      it("should modify 'to' field and add usa country code", async () => {
        process.argv = getCommandOptions({ from: "415935", to: "4159353327" });
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when command given invalid 'to' number", () => {
      it("should show error and usage", async () => {
        process.argv = getCommandOptions({ from: "4159353327", to: "415935" });
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });
  });

  describe("valid options", () => {
    describe("when command given help option", () => {
      it("should only display usage", async () => {
        process.argv = ["node", "prco-check-status", "-h"];
        getOptions().catch(async (e) => {
          expect(e).toMatchSnapshot();
        });
      });
    });

    describe("when comand given all valid options", () => {
      it("should return object containing the valid options", async () => {
        process.argv = ["node", "prco-check-status", "-t", "+14159353327", "-m", "booya"];
        response = await getOptions();
        response.twilioAccountSid = "hidden";
        response.twilioAuthToken = "hidden";
        response.signalwireSpaceUrl = "hidden";
        response.signalwireProjectId = "hidden";
        response.signalwireApiToken = "hidden";

        expect(response).toMatchSnapshot();
      });
    });
  });
});
