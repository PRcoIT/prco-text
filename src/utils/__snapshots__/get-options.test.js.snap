// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-options invalid options when comand given invalid 'to' option should show error and usage 1`] = `
Object {
  "error": [Error: Missing 'to' option.
],
  "errorMessage": "Missing 'to' option.


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "from": "+16505297726",
    "s": "signalwire",
    "service": "signalwire",
  },
}
`;

exports[`get-options invalid options when comand given invalid 'to' option should show error and usage 2`] = `
Object {
  "error": [Error: Missing 'message' option.
],
  "errorMessage": "Missing 'message' option.


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "from": "+16505297726",
    "s": "signalwire",
    "service": "signalwire",
    "t": " 555",
    "to": " 555",
  },
}
`;

exports[`get-options invalid options when command given invalid 'from' option should show error and usage 1`] = `
Object {
  "error": [Error: Missing 'to' option.
],
  "errorMessage": "Missing 'to' option.


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "from": "+16505297726",
    "s": "signalwire",
    "service": "signalwire",
  },
}
`;

exports[`get-options invalid options when command given invalid 'from' option should show error and usage 2`] = `
Object {
  "error": [Error: Missing 'to' option.
],
  "errorMessage": "Missing 'to' option.


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "f": " 555",
    "from": " 555",
    "s": "signalwire",
    "service": "signalwire",
  },
}
`;

exports[`get-options invalid options when command given invalid 'message' option should show error and usage 1`] = `
Object {
  "error": [Error: Missing 'to' option.
],
  "errorMessage": "Missing 'to' option.


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "from": "+16505297726",
    "s": "signalwire",
    "service": "signalwire",
  },
}
`;

exports[`get-options invalid options when command given invalid service option (not wis or oneguard) should show error and usage 1`] = `
Object {
  "error": [Error: Invalid service: use twilio or signalwire.
],
  "errorMessage": "Invalid service: use twilio or signalwire.


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "s": " foo",
    "service": " foo",
  },
}
`;

exports[`get-options invalid options when command given no options should produce error message and show usage 1`] = `
Object {
  "error": [Error: Missing 'to' option.
],
  "errorMessage": "Missing 'to' option.


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "from": "+16505297726",
    "s": "signalwire",
    "service": "signalwire",
  },
}
`;

exports[`get-options invalid options when config_env_file not found should show error and usage 1`] = `
Object {
  "error": [Error: Missing env file:  /foo],
  "errorMessage": "Missing env file:  /foo

USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {},
}
`;

exports[`get-options phone number transformation and validation when command given 'from' missing country code should modify 'from' field and add usa country code 1`] = `
Object {
  "error": [Error: Invalid 'to' field: invalid phone format: 415935
],
  "errorMessage": "Invalid 'to' field: invalid phone format: 415935


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "f": "4159353327",
    "from": "+14159353327",
    "m": "hi",
    "message": "hi",
    "s": "signalwire",
    "service": "signalwire",
    "t": "415935",
    "to": undefined,
  },
}
`;

exports[`get-options phone number transformation and validation when command given 'from' with dashes should modify 'from' field and remove dashes and add usa country code 1`] = `
Object {
  "error": [Error: Invalid 'to' field: invalid phone format: 415935
],
  "errorMessage": "Invalid 'to' field: invalid phone format: 415935


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "f": "415-935-3327",
    "from": "+14159353327",
    "m": "hi",
    "message": "hi",
    "s": "signalwire",
    "service": "signalwire",
    "t": "415935",
    "to": undefined,
  },
}
`;

exports[`get-options phone number transformation and validation when command given 'to' missing country code should modify 'to' field and add usa country code 1`] = `
Object {
  "error": [Error: Invalid 'from' field: invalid phone format: 415935
],
  "errorMessage": "Invalid 'from' field: invalid phone format: 415935


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "f": "415935",
    "from": undefined,
    "m": "hi",
    "message": "hi",
    "s": "signalwire",
    "service": "signalwire",
    "t": "4159353327",
    "to": "+14159353327",
  },
}
`;

exports[`get-options phone number transformation and validation when command given 'to' with dashes should modify 'to' field and remove dashes and add usa country code 1`] = `
Object {
  "error": [Error: Invalid 'from' field: invalid phone format: 415935
],
  "errorMessage": "Invalid 'from' field: invalid phone format: 415935


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "f": "415935",
    "from": undefined,
    "m": "hi",
    "message": "hi",
    "s": "signalwire",
    "service": "signalwire",
    "t": "415-935-3327",
    "to": "+14159353327",
  },
}
`;

exports[`get-options phone number transformation and validation when command given invalid 'from' number should show error and usage 1`] = `
Object {
  "error": [Error: Invalid 'from' field: invalid phone format: 4159353
],
  "errorMessage": "Invalid 'from' field: invalid phone format: 4159353


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "f": "4159353",
    "from": undefined,
    "m": "hi",
    "message": "hi",
    "s": "signalwire",
    "service": "signalwire",
    "t": "4159353327",
    "to": "+14159353327",
  },
}
`;

exports[`get-options phone number transformation and validation when command given invalid 'to' number should show error and usage 1`] = `
Object {
  "error": [Error: Invalid 'to' field: invalid phone format: 415935
],
  "errorMessage": "Invalid 'to' field: invalid phone format: 415935


USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "f": "4159353327",
    "from": "+14159353327",
    "m": "hi",
    "message": "hi",
    "s": "signalwire",
    "service": "signalwire",
    "t": "415935",
    "to": undefined,
  },
}
`;

exports[`get-options valid options when comand given all valid options should return object containing the valid options 1`] = `
Object {
  "from": "+16505297726",
  "message": "booya",
  "service": "signalwire",
  "signalwireApiToken": "hidden",
  "signalwireFrom": "+16505297726",
  "signalwireProjectId": "hidden",
  "signalwireSpaceUrl": "hidden",
  "to": "+14159353327",
  "twilioAccountSid": "hidden",
  "twilioAuthToken": "hidden",
  "twilioFrom": "+16507294319",
}
`;

exports[`get-options valid options when command given help option should only display usage 1`] = `
Object {
  "error": [Error],
  "errorMessage": "

USAGE

    prco-text [options]

    OPTIONS

        -c,--config_env_file  location of file containing environment variables
                              defaults to $HOME/protected/prco-text-env
        -s,--service          twilio or signalwire -- default: signalwire
        -f,--from             originating phone number -- default: number in .env file
        -t,--to               destination phone number
        -m,--message          message to be sent
        -h,--help             display usage help

EXAMPLE

    $ prco-text --to 415-222-3333 --message 'hello,
    Your car is ready :)

    PRCO
    415-555-1212'


  ",
  "options": Object {
    "_": Array [],
    "h": "",
    "help": "",
    "s": "signalwire",
    "service": "signalwire",
  },
}
`;
